Compile tests for multiple packages

fhahn opened the pull request:
This PR for #1528 is still pretty rough (and so far only added multiple package support for tests), but I wanted to make sure the overall approach is fine.

I have made some progress and cargo is now able to compile and execute tests for multiple packages. 

In order to execute the doc tests for multiple packages as well, I added `to_doc_test` to `Compilation` (used [here](https://github.com/fhahn/cargo/commit/8b8df3e8d5abfa638852f55730d2f3cb4eec66bc#diff-417e085367d0ce027505dfaa26a089d3R33)). Previously it only executed the doc tests for `Compilation.package`. Another option would be to change `Compilation.package` to be a `Vec<Package>`, because in order to execute the tests of multiple packages, multiple packages are compiled as "top level package". But this would require a changes at a couple of other places.

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented on src/bin/test.rs:
```diff
@@ -26,7 +26,7 @@ pub const USAGE: &'static str = "
 
 Usage:
-    cargo test [options] [--] [<args>...]
+    cargo test [options] [-p SPEC --package SPEC]... [--] [<args>...]
```
cc @BurntSushi, you mentioned on https://github.com/docopt/docopt/issues/275 that you were thinking of proposing the ability to do this in the options section instead of up here in the usage section, do you know if there's been progress on that? I wouldn't mind proposing an idea or two on that thread to see if I can get discussion rolling again, it'd be great to fix )

BurntSushi replied:
@alexcrichton Thanks for the reminder! I have a candidate implementation working in `0.6.69`. See my proposal for details: https://github.com/docopt/docopt/issues/275 (Even though it's working in the Rust implementation, I probably wouldn't add it into Cargo quite yet in case the proposal changes. If it's ultimately rejected, I might still leave it in because I think it's really really useful.)

alexcrichton replied:
Sounds good to me, thanks @BurntSushi!

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -101,11 +101,58 @@ pub fn compile<'a>(manifest_path: &Path,
+fn resolve_dependencies<'a>(options: &CompileOptions<'a>,
+                 package: &Package,
+                 source: Option<Box<Source + 'a>>
+                 ) -> (Vec<Package>, Resolve, SourceMap<'a>) {
```
Stylistically these kinds of functions are formatted as this in cargo:

```rust
fn foo(a: A,
       b: B)
       -> Ret {
    // ...
```

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -101,11 +101,58 @@ pub fn compile<'a>(manifest_path: &Path,
+                 package: &Package,
+                 source: Option<Box<Source + 'a>>
+                 ) -> (Vec<Package>, Resolve, SourceMap<'a>) {
+    let override_ids = source_ids_from_config(options.config, package.root()).unwrap();
```
All the `unwrap()` calls in this function should be converted to `try!`

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -123,78 +170,52 @@ pub fn compile_pkg<'a>(package: &Package,
+    };
+    let pkgid = pkgids.first().unwrap();
+    let to_build = packages.iter().find(|p| p.package_id() == *pkgid).unwrap();
+    let to_builds = packages.iter().filter(|p| pkgids.iter().find(|&op| *op == p.package_id()).is_some()).collect::<Vec<&Package>>();
```
Can you be sure that lines are word-wrapped to 80 characters in Cargo?

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -95,36 +97,32 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
 
     let mut queue = JobQueue::new(cx.resolve, deps, cx.jobs());
 
+    let _p = profile::start("preparing build directories");
```
This ends up finishing the profiling when it's dropped, which is now at the end of the function (which is why this wanted an explicit scope)

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -95,36 +97,32 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
     // Build up a list of pending jobs, each of which represent compiling a
     // particular package. No actual work is executed as part of this, that's
     // all done next as part of the `execute` function which will run
     // everything in order with proper parallelism.
```
This comment is now dangling.

alexcrichton commented:
Yeah this all looks like a pretty good approach to me!

bors commented:
:umbrella: The latest upstream changes (presumably #1830) made this pull request unmergeable. Please resolve the merge conflicts.

fhahn commented:
I finally had time to work on this PR again. I have started adding tests, but I am not sure how to test `cargo test` and `cargo bench`, because the output order is non deterministic for multiple packages (e.g. [test_cargo_test::test_multiple_packages](https://github.com/fhahn/cargo/commit/3212778fa4ebe72022a992b9cbb7affa2341b134#diff-c962b2c6409c073bd126f83278d97f78R1859)) Any ideas whats a good way to test this would be?

bors commented:
:umbrella: The latest upstream changes (presumably #1973) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
The nondeterminism may be solved by `-j1`, but if that doesn't work then just testing that they're accepted should be fine.

alexcrichton commented on src/bin/bench.rs:
```diff
@@ -26,7 +26,7 @@ pub const USAGE: &'static str = "
 
 Usage:
-    cargo bench [options] [--] [<args>...]
+    cargo bench [options] [-p SPEC --package SPEC]... [--] [<args>...]
```
I think that new syntax in docopt may actually allow for nicer definitions of this, something like:

```
Options:
    -p, --package=SPEC ...         Description ...
```

I believe it's still somewhat experimental, but this may be a good place to get a good start on it?

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -48,7 +48,7 @@ pub struct CompileOptions<'a> {
     pub no_default_features: bool,
     /// Root package to build (if None it's the current one)
-    pub spec: Option<&'a str>,
+    pub spec: Vec<&'a str>,
```
This may be much easier to take if it's a `&'a [String]`, that way all the commands above can avoid casting and collection (e.g. they're just `&options.flag_package`

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -123,15 +123,15 @@ pub fn compile_pkg<'a>(package: &Package,
         } else {
-            try!(registry.add_sources(&[package.package_id().source_id()
-                                               .clone()]));
+            try!(registry.add_sources(&[package.package_id().source_id().clone()]));
```
Were the changes around here necessary? (looks to be pretty similar)

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -145,10 +145,10 @@ pub fn compile_pkg<'a>(package: &Package,
             dev_deps: true, // TODO: remove this option?
             features: &features,
-            uses_default_features: !no_default_features,
+            uses_default_features: !options.no_default_features,
```
Same as above, probably good to cut down to as few changes as possible.

alexcrichton replied:
(ping on this)

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -165,33 +165,66 @@ pub fn compile_pkg<'a>(package: &Package,
+            id
+        }).filter(|id|
+            id.is_ok()
+        ).map(|id| id.unwrap()).collect::<Vec<&PackageId>>()
```
This may be better expressed as:

```rust
spec.iter().filter_map(|p| {
    match resolved_with_overrides.query(p) {
        Ok(p) => Some(p),
        Err(..) => { invalid_spec.push(p); None }
    }
}).collect::<Vec<_>>();
```

I'm also a little hesitant about throwing away the error information, I think invalid specifications will want to retain the error information about *why* they were invalid

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -165,33 +165,66 @@ pub fn compile_pkg<'a>(package: &Package,
+    }
+
+    let pkgid = pkgids[0];
     let to_build = packages.iter().find(|p| p.package_id() == pkgid).unwrap();
```
Isn't this variable obsolete now? There could be more than one package being built?

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -54,8 +54,9 @@ pub struct TargetConfig {
 // where the compiled libraries are all located.
-pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
-                                     pkg: &'a Package,
+pub fn compile_targets<'a, 'cfg: 'a>(pkg_targets: &'a Vec<(&Package,
```
`&'a Vec<T>` should basically always be avoided in favor of `&'a [T]` explicitly.

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -63,7 +64,9 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
-    if targets.is_empty() {
+
+    let &(pkg, _) = pkg_targets.first().unwrap();
+    if pkg_targets.iter().any(|&(_, ref targets)| targets.is_empty()) {
```
This isn't quite right any more I think because singling out the first package isn't really possible any more. This entire `if` may actually just be able to go away.

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -72,37 +75,36 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
-    } else {
-        deps.iter().find(|p| p.package_id() == resolve.root()).unwrap()
-    };
+    let root = deps.iter().find(|p| p.package_id() == resolve.root()).unwrap();
```
I think this logic will need to be updated, the notion of a "root" package doesn't really exist any more, so those which are depending on it probably need to be updated.

fhahn replied:
At the moment `-p SPEC` only builds packages that are dependencies of the "root" package, and all generated binaries are stored relative to that root package. I think this root package is mostly used to get a base path for the build output and I think this notation should still be valid?

alexcrichton replied:
Hm the placement of the output directory should be relatively independent from the root package (e.g. it's just a default), as in once the output directory is chosen the root package is never needed after that. 

I'll try to audit for usage of this regardless, though, to see what's up.

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -72,37 +75,36 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
-    // Prep the context's build requirements and see the job graph for all
-    // packages initially.
     {
         let _p = profile::start("preparing build directories");
```
This profile marker isn't quite profiling the same thing any more, could you be sure to keep the scope the same as it was before?

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -110,42 +112,47 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
-                cx.compilation.libraries.insert(pkgid.clone(), v);
+
+                // Include immediate lib deps as well
+                for dep in cx.dep_targets(pkg, target, kind, profile).iter() {
```
This addition of `.iter()` I don't think is necessary

alexcrichton commented on src/cargo/ops/cargo_test.rs:
```diff
@@ -30,16 +30,19 @@ pub fn run_tests(manifest_path: &Path,
         return Ok(None)
     }
 
-    let libs = compile.package.targets().iter()
```
I think this field in `Compilation`, `package`, should probably be removed now as it's not quite as correct as it was before.

alexcrichton commented:
cc @brson, this is touching quite a few Cargo internals

bors commented:
:umbrella: The latest upstream changes (presumably #1949) made this pull request unmergeable. Please resolve the merge conflicts.

fhahn commented:
I hope I managed to address all of @alexcrichton's feedback with the update. I also added support for multiple `-p`options to `clean` and `update`.

However I noticed that `cargo update` seems to ignore the `-p` option. I have added a [test](https://github.com/rust-lang/cargo/pull/1828/files#diff-ccf7349a39a2e472790c5235e8cd9adaR151) for this, but I could also open an issue?

alexcrichton commented on src/cargo/ops/cargo_clean.rs:
```diff
@@ -25,48 +25,51 @@ pub fn clean(manifest_path: &Path, opts: &CleanOptions) -> CargoResult<()> {
-    };
 
-    // Create a compilation context to have access to information like target
+     // Create a compilation context to have access to information like target
```
stray space

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -101,11 +101,11 @@ pub fn compile<'a>(manifest_path: &Path,
                        options: &CompileOptions<'a>)
                        -> CargoResult<ops::Compilation<'a>> {
-    let CompileOptions { config, jobs, target, spec, features,
+    let CompileOptions { config, jobs, target, ref spec, features,
```
I think this `ref` can be omitted (it's just a slice)

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -165,33 +165,62 @@ pub fn compile_pkg<'a>(package: &Package,
+    if spec.len() > 0 && invalid_spec.len() > 0 {
+        return Err(human(format!("could not find package matching spec `{}`",
+                                 invalid_spec.join(", "))));
+    } */
```
Should be uncommented? May be worth ensuring there's a test for this as well

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -165,33 +165,62 @@ pub fn compile_pkg<'a>(package: &Package,
+
+    let to_builds = packages.iter().filter(|p|
+        pkgids.iter().find(|&op| *op == p.package_id()).is_some()
+    ).collect::<Vec<&Package>>();
```
Conventionally in Cargo multi-line closures are surrounded by `{` braces

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -165,33 +165,62 @@ pub fn compile_pkg<'a>(package: &Package,
+    } */
+
+    let to_builds = packages.iter().filter(|p|
+        pkgids.iter().find(|&op| *op == p.package_id()).is_some()
```
Could this be `pkgids.contains(p.package_id())`?

alexcrichton commented on src/bin/rustc.rs:
```diff
@@ -32,22 +32,22 @@ Usage:
-    -q, --quiet              No output printed to stdout
-    --color WHEN             Coloring: auto, always, never
+    -h, --help                   Print this message
+    -p SPEC, --package SPEC ...  The profile to compile for
```
This is actually a command where it's always an error to specify more than one package, so it's ok to leave this as it was.

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -165,33 +165,62 @@ pub fn compile_pkg<'a>(package: &Package,
+        pkgids.iter().find(|&op| *op == p.package_id()).is_some()
+    ).collect::<Vec<&Package>>();
+
+    let mut twas = &mut vec![];
```
Could this get a more descriptive variable name? Also the `&mut` in front should be fine to omit and  conventionally Cargo uses `Vec::new()` instead of `vec![]`

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -165,33 +165,62 @@ pub fn compile_pkg<'a>(package: &Package,
+    }
+
+    for targets in twas {
+        let (target, ref profile) = *targets;
```
This destructuring could be done in the `for` loop pattern binding perhaps

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -165,33 +165,62 @@ pub fn compile_pkg<'a>(package: &Package,
+        let (target, ref profile) = *targets;
+        for &to_build in to_builds.iter() {
+            package_targets.push((to_build, vec![(target, profile)]));
         }
```
This logic is a little confusing for me, I think this is only triggered in the case of where `target_rustc_args` is passed, but I think it may be good to keep that logic outside these loops as in that scenario you're required to only have one target anyway.

alexcrichton commented on src/cargo/ops/cargo_doc.rs:
```diff
@@ -44,13 +44,15 @@ pub fn doc(manifest_path: &Path,
-                    None => return Ok(())
-                }
+        let name = if options.compile_opts.spec.len() > 0{
+            // TODO
```
Could you have this generate an error if `--open` is passed and multiple packages are specified?

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -63,89 +65,97 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
 
-    debug!("compile_targets: {}", pkg);
+    debug!("compile_targets: {}", pkg_targets.iter().map(|&(ref p, _)| p.name())
+                                                    .collect::<Vec<_>>().connect(", "));
```
It's ok to just drop this debug statement, it never really bought much anyway

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -63,89 +65,97 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
+        // packages initially.
+        for &(pkg, ref targets) in pkg_targets {
+            try!(cx.prepare(pkg, targets));
+            prepare_init(&mut cx, pkg, &mut queue, &mut HashSet::new());
```
I think this may want to use the same visited hash set among all calls to this function, that'll prevent re-preparing the same package twice.

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -63,89 +65,97 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
+        for &(pkg, ref targets) in pkg_targets {
+            try!(cx.prepare(pkg, targets));
+            prepare_init(&mut cx, pkg, &mut queue, &mut HashSet::new());
+            custom_build::build_map(&mut cx, pkg, targets);
```
This function isn't currently really designed to be called more than once, and actually taking a look at these... most of them shouldn't actually work after being called more than once. I think the passing of multiple targets will need to be threaded a bit deeper to each of these so the loop can happen farther down rather than up here.

It'd be fine to define a type alias for this package of "things to build" if typing out the slice and types gets too verbose.

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -199,18 +209,17 @@ fn compile<'a, 'cfg>(targets: &[(&'a Target, &'a Profile)],
+                (false, false, _) if !target.is_bin() => Stage::BinaryTests,
+                (false, false, _) => Stage::Binaries,
             };
+            let dst = jobs.queue(pkg, stage);
```
I don't think any logic changed here, right? Could this stay the same as it was before?

alexcrichton commented on src/cargo/ops/cargo_test.rs:
```diff
@@ -69,7 +69,10 @@ fn compile_tests<'a>(manifest_path: &Path,
+    compilation.tests.iter_mut()
+                     .map(|&mut (_, ref mut tests)|
+                          tests.sort_by(|&(ref n1, _), &(ref n2, _)| n1.cmp(n2)))
+                     .collect::<Vec<_>>();
```
Perhaps this could just be:

```rust
for &mut (_, ref mut tests) in compilation.tests.iter_mut() {
    tests.sort();
}
```

alexcrichton replied:
Er actually:

```rust
for tests in compilation.tests.iter_mut() {
    tests.1.sort();
}
```

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -63,89 +65,97 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
-                cx.compilation.libraries.insert(pkgid.clone(), v);
+                    let kind = kind.for_target(target);
+                    let v =
+                        try!(cx.target_filenames(pkg, target, profile, kind));
```
Line breaking in Cargo typically happens on arguments instead of after `=` signs like this

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -63,89 +65,97 @@ pub fn compile_targets<'a, 'cfg: 'a>(targets: &[(&'a Target, &'a Profile)],
+        cx.compilation.tests.push((pkg.clone(), tests));
+
+        if let Some(feats) = cx.resolve.features(pkg.package_id()) {
+            cx.compilation.features.extend(feats.iter().cloned());
```
I think that this should only happen for the "root" package as that's the only one you're allowed to modify the features of (e.g. with multiple packages you can't modify features)

alexcrichton commented on tests/test_cargo_rustc.rs:
```diff
@@ -296,3 +298,68 @@ test!(build_only_bar_dependency {
+        "#);
+    baz.build();
+
+    assert_that(foo.cargo_process("rustc").arg("-v").arg("-p").arg("bar")
```
This can actually be just `foo.cargo` as the project was already built previously.

Also, this should actually generate an error as you can't specify more than one target to `cargo rustc` currently.

alexcrichton commented:
Very nice work, thanks for the thorough tests @fhahn!

I'm a little surprised to hear that `cargo update` would ignore `-p`, I thought there were tests previously ensuring that worked... Are you sure it wasn't accidentally introduced by this change?

fhahn commented:
If pushed a commit to address @alexcrichton's  helpful feedback, thanks. I also made changes how `cargo update` handles multiple packages.

And I think the failing test I added does not make much sense (the test only checked to lockfile), I added another one which should really test `cargo update` with multiple packages

alexcrichton commented on src/cargo/core/registry.rs:
```diff
@@ -174,14 +177,18 @@ impl<'cfg> PackageRegistry<'cfg> {
 
-    fn load(&mut self, source_id: &SourceId, kind: Kind) -> CargoResult<()> {
+    fn load(&mut self, source_id: &SourceId, kind: Kind, 
+            updated: &mut HashSet<String>) -> CargoResult<()> {
```
How come this parameter was added? There's other machinery in play here to ensure that we don't update sources twice, so I'm curious what triggered this path.

fhahn replied:
Do you by any chance know where this machinery lives? I would be happy to reuse it.

Without this set, the registry is updated multiple times, if multiple packages are passed to `cargo update` ([test case](https://github.com/rust-lang/cargo/pull/1828/files#diff-fcf81cfe756ff54027fad304503760d2R833)), because the souces array passed to `add_sources` [here](https://github.com/rust-lang/cargo/pull/1828/files#diff-72d3ed8b640540c0f99dbe2877ef0ad6R84) can contain the same source (e.g. the registry) multiple times.

alexcrichton commented on src/cargo/ops/cargo_rustc/context.rs:
```diff
@@ -143,36 +143,40 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
             internal(format!("couldn't prepare build directories for `{}`",
-                             pkg.name()))
+                             pkgs.iter().map(|&(ref pkg, _)| format!("`{}`", pkg.name()))
+                                        .collect::<Vec<_>>().join(",")))
```
It's ok to just omit the names of the packages here, "couldn't prepare build directories" should be good enough (same below as well)

alexcrichton commented on src/cargo/ops/cargo_test.rs:
```diff
@@ -114,15 +119,20 @@ fn run_doc_tests(options: &TestOptions,
-                    .map(|t| (t.src_path(), t.name(), t.crate_name()));
-    for (lib, name, crate_name) in libs {
+
+    let mut libs = vec![];
```
You may be able to avoid the intermediate vec here (mostly for clarity) via a `flat_map` instead of a vec + extend

fhahn replied:
Thanks, I've changed the code. I did not manage to do it with `flat_map`, because the borrow checker complained that `package` does not live long enough (when used in `map`):

     let libs = compilation.to_doc_test.iter().flat_map(|package| {
         package.targets().iter().filter(|t| t.doctested())
                          .map(|t| (package, t.src_path(), t.name(), t.crate_name()))
     }).collect::<Vec<_>>();

Now libs is an iterator over `(&Package, Iterator<(lib, name, create_name))`

alexcrichton commented:
Looking good! There's some failing tests still and a few minor nits, but otherwise this is looking good to go

bors commented:
:umbrella: The latest upstream changes (presumably #1991) made this pull request unmergeable. Please resolve the merge conflicts.

fhahn commented:
The test failures shoud have been due to using `to_url` [here](https://github.com/fhahn/cargo/commit/d1cd5ae19c2caca9f99bacea577bbdd57d658a84#diff-2849d70734c6b3e65862ff4190fd1abaR185), which now uses the `SourceId` directly.

alexcrichton commented on src/cargo/core/registry.rs:
```diff
@@ -143,20 +143,23 @@ impl<'cfg> PackageRegistry<'cfg> {
+        let mut updated = HashSet::new();
         for id in ids.iter() {
-            try!(self.load(id, Kind::Locked));
+            try!(self.load(id, Kind::Locked, &mut updated));
```
You may want to try using `ensure_loaded` here instead of `load` as that should handle the deduplication for you

alexcrichton commented:
Ah ok, perhaps the use of `ensure_loaded` instead of `load` in `add_sources` will fix the tets?

fhahn commented:
Using `ensure_loaded` worked, I just had to pass the correct `Kind` along as a parameter, so `Kind::Locked` is passed anlong to `load` from `add_sources`

alexcrichton commented:
Looks good to me, thanks @fhahn! I believe the Windows failures in the CI are legit, however (tests needing a `.exe` at the end when probing for a file)

alexcrichton commented:
Ok, I think those failures on CI are spurious (only seen 1/3 before...), but could you also squash these commits down? It's ok to have a few, but I'd prefer to avoid "Address @alexcrichton's feedback" commits if possible :)

fhahn commented:
I've squashed the commits down to 6, but I could reduce the number even further if necessary.

alexcrichton commented:
Looks like this may have accidentally picked up some other commits, perhaps a rebase is in order? I'll give this one last look-through but I'll probably just send it to bors.

alexcrichton commented:
Now that I've actually gotten around to taking a peek, this looks fantastic to me. Thanks so much again for pushing on this @fhahn!

@bors: r+ d9615d70f311ad80fba943fbb5c3df477786d9c9

bors commented:
:hourglass: Testing commit d9615d7 with merge 7931e58...

bors commented:
:sunny: Test successful - [cargo-linux-32](http://buildbot.rust-lang.org/builders/cargo-linux-32/builds/1317), [cargo-linux-64](http://buildbot.rust-lang.org/builders/cargo-linux-64/builds/1306), [cargo-mac-32](http://buildbot.rust-lang.org/builders/cargo-mac-32/builds/1313), [cargo-mac-64](http://buildbot.rust-lang.org/builders/cargo-mac-64/builds/1310), [cargo-win-gnu-32](http://buildbot.rust-lang.org/builders/cargo-win-gnu-32/builds/189), [cargo-win-gnu-64](http://buildbot.rust-lang.org/builders/cargo-win-gnu-64/builds/190), [cargo-win-msvc-32](http://buildbot.rust-lang.org/builders/cargo-win-msvc-32/builds/110), [cargo-win-msvc-64](http://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/136)

fhahn commented:
Great, thanks @alexcrichton for your patience :)
